aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp

```diff
@@ -109,6 +109,251 @@ void inline col_indices_and_values_resize_(const Tensor& input, int64_t nnz) {
+            block_size,
+            info.descriptor(),
+            &buffer_size);
+
+        auto& allocator = *c10::cuda::CUDACachingAllocator::get();
+        auto work_data = allocator.allocate(buffer_size);
+
+        at::cuda::sparse::bsrsv2_analysis(
+            handle,
+            block_layout,
+            opA,
+            mb,
+            nnzb,
+            desc.descriptor(),
+            values_data_ptr,
+            crow_indices_data_ptr,
+            col_indices_data_ptr,
+            block_size,
+            info.descriptor(),
+            CUSPARSE_SOLVE_POLICY_NO_LEVEL,
+            work_data.get());
+
+        at::cuda::sparse::bsrsv2_solve(
+            handle,
+            block_layout,
+            opA,
+            mb,
+            nnzb,
+            &alpha,
+            desc.descriptor(),
+            values_data_ptr,
+            crow_indices_data_ptr,
+            col_indices_data_ptr,
+            block_size,
+            info.descriptor(),
+            B_->data_ptr<scalar_t>(),
+            X_->data_ptr<scalar_t>(),
+            CUSPARSE_SOLVE_POLICY_NO_LEVEL,
+            work_data.get());
+      });
+  if (!X.is_same(*X_)) {
+    X.copy_(*X_);
+  }
+}
+
+void block_sparse_triangular_solve_mat(
+    const at::sparse_csr::SparseCsrTensor& A,
+    const Tensor& B,
+    const Tensor& X,
+    bool upper,
+    bool transpose,
+    bool unitriangular) {
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(A.is_sparse_csr());
+  // values is expected to be a blocks of sparse matrix
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(A.values().dim() == 3);
+  // blocks are expected to be square
+  TORCH_INTERNAL_ASSERT(A.values().size(2) == A.values().size(1));
+  // only block of size > 1 is supported in cuSPARSE
+  TORCH_INTERNAL_ASSERT(A.values().size(-1) > 1);
+  // blocks are expected to be in row- or column-major order
+  TORCH_INTERNAL_ASSERT(
+      A.values().is_contiguous() ||
+      A.values().transpose(-2, -1).is_contiguous());
+
+  // cuSPARSE can't work with empty sparse matrices
+  if (A._nnz() == 0) {
+    X.fill_(NAN);
+    return;
+  }
+
+  const cusparseDirection_t block_layout = A.values().is_contiguous()
+      ? CUSPARSE_DIRECTION_ROW
+      : CUSPARSE_DIRECTION_COLUMN;
+
+  // c10::MaybeOwned<Tensor> X_ = prepare_dense_matrix_for_cusparse(X);
+  // // cuSPARSE doesn't let us pass transpose flag separately for X and B
+  // // therefore we use same strides for both X and B
+  // // if X is row-major, we set the transpose flag for B and X
+  // c10::MaybeOwned<Tensor> B_ =
+  //     prepare_dense_matrix_for_cusparse(B, X_->strides());
+
+  c10::MaybeOwned<Tensor> X_ = prepare_column_major_matrix_for_cusparse(X);
+  c10::MaybeOwned<Tensor> B_ = prepare_column_major_matrix_for_cusparse(B);
+
+  IntArrayRef X_strides = X_->strides();
+  auto ndim = X_->dim();
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(ndim == 2);
+  bool is_X_row_major = (X_strides[ndim - 1] == 1);
+  // int ldx = is_X_row_major ? cuda_int_cast(X_strides[ndim - 2], "ldx")
```

Comment:
nit: Did you mean to remove this?

Reply:
Yes, I need to clean up this and there is another commented code is left in a few places.